ray-operator/controllers/raycluster_controller_fake_test.go

```diff
@@ -0,0 +1,373 @@
+/*
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	. "github.com/onsi/ginkgo"
+	rayiov1alpha1 "github.com/ray-project/kuberay/ray-operator/api/raycluster/v1alpha1"
+	"github.com/ray-project/kuberay/ray-operator/controllers/common"
+	"github.com/ray-project/kuberay/ray-operator/pkg/client/clientset/versioned/scheme"
+	"github.com/stretchr/testify/assert"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	clientFake "sigs.k8s.io/controller-runtime/pkg/client/fake"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/log/zap"
+	"testing"
+
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/utils/pointer"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	// +kubebuilder:scaffold:imports
+)
+
+var (
+	namespaceStr     string
+	instanceName     string
+	groupNameStr     string
+	expectReplicaNum int32
+	testPods         []runtime.Object
+	testRayCluster   *rayiov1alpha1.RayCluster
+)
+
+func setupTest() {
+	logf.SetLogger(zap.New(zap.WriteTo(GinkgoWriter), zap.UseDevMode(true)))
+
+	namespaceStr = "default"
+	instanceName = "raycluster-sample"
+	groupNameStr = "small-group"
+	expectReplicaNum = 3
+	testPods = []runtime.Object{
+		&corev1.Pod{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "headNode",
+				Namespace: namespaceStr,
+				Labels: map[string]string{
+					common.RayClusterLabelKey:   instanceName,
+					common.RayNodeTypeLabelKey:  string(rayiov1alpha1.HeadNode),
+					common.RayNodeGroupLabelKey: groupNameStr,
```

Comment:
I think this should be "headgroup", not "small-group". The reconcilePods method does filter to get headPods and workerPods. Does your setup reproduce that correctly?

Reply:
Good point. I will update the group name. This test setup just considers a part of pod resources.